### PR TITLE
Allow to load multiple animation/libraries at once in the animation manager

### PR DESCRIPTION
**Pull request by YeldhamDev**
Currently, you can only load animations and animation libraries one-by-one in the "Edit Animation Libraries" dialog. This PR allows the user to now pick multiple ones at once to load.

### CODE REVIEW

**Review by SaracenOne — APPROVED** [review]
Tested, big UX improvement, LGTM

**Review by SaracenOne — CHANGES_REQUESTED** [review]
Could you implement the changes @AThousandShips requested please?

**Comment by KoBeWi** [comment]
Some crash when loading multiple (empty) AnimationLibraries:
```
CrashHandlerException: Program crashed
Engine version: Godot Engine v4.2.beta.custom_build (5ee983188de97ae027f9b9c1443438063f708a7e)
Dumping the backtrace. Please include this when reporting the bug to the project developer.
[0] VectorWriteProxy<Variant>::operator[] (C:\godot_source\core\templates\vector.h:52)
[1] Array::operator[] (C:\godot_source\core\variant\array.cpp:90)
[2] AnimationLibraryEditor::_load_files (C:\godot_source\editor\plugins\animation_library_editor.cpp:364)
[3] call_with_variant_args_helper<AnimationLibraryEditor,Vector<String>,0> (C:\godot_source\core\variant\binder_common.h:308)
[4] call_with_variant_args<AnimationLibraryEditor,Vector<String> > (C:\godot_source\core\variant\binder_common.h:418)
[5] CallableCustomMethodPointer<AnimationLibraryEditor,Vector<String> >::call (C:\godot_source\core\object\callable_method_pointer.h:105)
[6] Callable::callp (C:\godot_source\core\variant\callable.cpp:58)
[7] Object::emit_signalp (C:\godot_source\core\object\object.cpp:1127)
[8] Node::emit_signalp (C:\godot_source\scene\main\node.cpp:3608)
[9] Object::emit_signal<Vector<String> > (C:\godot_source\core\object\object.h:920)
[10] EditorFileDialog::_action_pressed (C:\godot_source\editor\gui\editor_file_dialog.cpp:398)
[11] call_with_variant_args_helper<EditorFileDialog> (C:\godot_source\core\variant\binder_common.h:308)
[12] call_with_variant_args<EditorFileDialog> (C:\godot_source\core\variant\binder_common.h:418)
[13] CallableCustomMethodPointer<EditorFileDialog>::call (C:\godot_source\core\object\callable_method_pointer.h:105)
[14] Callable::callp (C:\godot_source\core\variant\callable.cpp:58)
[15] Object::emit_signalp (C:\godot_source\core\object\object.cpp:1127)
[16] Node::emit_signalp (C:\godot_source\scene\main\node.cpp:3608)
[17] Object::emit_signal<> (C:\godot_source\core\object\object.h:920)
[18] AcceptDialog::_ok_pressed (C:\godot_source\scene\gui\dialogs.cpp:121)
[19] call_with_variant_args_helper<AcceptDialog> (C:\godot_source\core\variant\binder_common.h:308)
[20] call_with_variant_args<AcceptDialog> (C:\godot_source\core\variant\binder_common.h:418)
[21] CallableCustomMethodPointer<AcceptDialog>::call (C:\godot_source\core\object\callable_method_pointer.h:105)
[22] Callable::callp (C:\godot_source\core\variant\callable.cpp:58)
[23] Object::emit_signalp (C:\godot_source\core\object\object.cpp:1127)
[24] Node::emit_signalp (C:\godot_source\scene\main\node.cpp:3608)
[25] Object::emit_signal<> (C:\godot_source\core\object\object.h:920)
[26] BaseButton::_pressed (C:\godot_source\scene\gui\base_button.cpp:139)
[27] BaseButton::on_action_event (C:\godot_source\scene\gui\base_button.cpp:177)
[28] BaseButton::gui_input (C:\godot_source\scene\gui\base_button.cpp:70)
[29] Control::_call_gui_input (C:\godot_source\scene\gui\control.cpp:1810)
[30] Viewport::_gui_call_input (C:\godot_source\scene\main\viewport.cpp:1596)
[31] Viewport::_gui_input_event (C:\godot_source\scene\main\viewport.cpp:1862)
[32] Viewport::push_input (C:\godot_source\scene\main\viewport.cpp:3186)
[33] Window::_window_input (C:\godot_source\scene\main\window.cpp:1553)
[34] call_with_variant_args_helper<Window,Ref<InputEvent> const &,0> (C:\godot_source\core\variant\binder_common.h:303)
[35] call_with_variant_args<Window,Ref<InputEvent> const &> (C:\godot_source\core\variant\binder_common.h:418)
[36] CallableCustomMethodPointer<Window,Ref<InputEvent> const &>::call (C:\godot_source\core\object\callable_method_pointer.h:105)
[37] Callable::callp (C:\godot_source\core\variant\callable.cpp:58)
[38] Callable::call<Ref<InputEvent> > (C:\godot_source\core\variant\variant.h:847)
[39] DisplayServerWindows::_dispatch_input_event (C:\godot_source\platform\windows\display_server_windows.cpp:2714)
[40] DisplayServerWindows::_dispatch_input_events (C:\godot_source\platform\windows\display_server_windows.cpp:2684)
[41] Input::_parse_input_event_impl (C:\godot_source\core\input\input.cpp:750)
[42] Input::flush_buffered_events (C:\godot_source\core\input\input.cpp:1012)
[43] DisplayServerWindows::process_events (C:\godot_source\platform\windows\display_server_windows.cpp:2399)
[44] OS_Windows::run (C:\godot_source\platform\windows\os_windows.cpp:1474)
[45] widechar_main (C:\godot_source\platform\windows\godot_windows.cpp:182)
[46] _main (C:\godot_source\platform\windows\godot_windows.cpp:204)
[47] main (C:\godot_source\platform\windows\godot_windows.cpp:218)
[48] WinMain (C:\godot_source\platform\windows\godot_windows.cpp:232)
[49] __scrt_common_main_seh (D:\a\_work\1\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:288)
[50] <couldn't map PC to fn name>
-- END OF BACKTRACE --
```

**Comment by YeldhamDev** [comment]
@SaracenOne @AThousandShips Changes made.

@KoBeWi How many libraries are you loading? I tried to reproduce the crash on my end, but no success.

**Comment by KoBeWi** [comment]
3.
I used Create New Resource to make libraries (empty) and loaded them in a newly created AnimationPlayer.

**Comment by YeldhamDev** [comment]
> 3.

Welp, I loaded 10, and no crash.

**Comment by KoBeWi** [comment]
https://github.com/godotengine/godot/assets/2223172/0759aa5f-ced9-456e-82a2-4d9005f21972

Seems like you need to add an animation before loading libraries.

**Comment by YeldhamDev** [comment]
@KoBeWi So, I found the problem... I mixed up the `i` and `j` in the `for` loop. :grimacing:

**Review by SaracenOne — APPROVED** [review]
Okay, LGTM! 👍

File: editor/plugins/animation_library_editor.cpp [hunk 1]
```diff
@@ -414,6 +338,103 @@ void AnimationLibraryEditor::_load_file(String p_path) {
...
+
+void AnimationLibraryEditor::_load_files(const PackedStringArray p_paths) {
+	EditorUndoRedoManager *undo_redo = EditorUndoRedoManager::get_singleton();
+	bool has_created_action = false;
+	bool show_error_diag = false;
+	List<String> anim_to_add;
+
+	switch (file_dialog_action) {
+		case FILE_DIALOG_ACTION_OPEN_LIBRARY: {
+			for (int i = 0; i < p_paths.size(); i++) {
```
Comment (KoBeWi) [inline editor/plugins/animation_library_editor.cpp:354]:
```suggestion
			for (const String &path : p_paths) {
```

File: editor/plugins/animation_library_editor.cpp [hunk 2]
```diff
@@ -414,6 +338,103 @@ void AnimationLibraryEditor::_load_file(String p_path) {
...
+				Ref<AnimationLibrary> al = ResourceLoader::load(p_paths[i]);
+				if (al.is_null()) {
+					show_error_diag = true;
+					error_dialog->set_text(TTR("Some AnimationLibrary files were invalid."));
+					continue;
+				}
+
+				TypedArray<StringName> libs = mixer->call("get_animation_library_list");
+				for (int j = 0; j < libs.size(); j++) {
+					const StringName K = libs[j];
```
Comment (KoBeWi) [inline editor/plugins/animation_library_editor.cpp:364]:
```suggestion
					const StringName &K = libs[j];
```

Comment (KoBeWi) [reply editor/plugins/animation_library_editor.cpp:364]: I just noticed this comes from the previous code.
#84551 fixes it.

**Comment by Illauriel** [comment]
Oh hey! I've actually made a similar PR recently, because this one didn't pop up in the search as it doesn't mention Animation Library Editor. 
- #84088

It can probably be closed now, but there's also another small QOL improvement, (re)enabling the tree fold arrow in the same screen, because scrolling through hundreds of animations is a PITA. Can be cherrypicked from second commit. (it's one line)

- This also Resolves godotengine/godot-proposals#8226

**Comment by YeldhamDev** [comment]
@Illauriel

> Oh hey! I've actually made a similar PR recently, because this one didn't pop up in the search as it doesn't mention Animation Library Editor.

It happens, don't worry. However, in your PR, each undo/redo action is done separately, instead of being contained into a single one.

> It can probably be closed now, but there's also another small QOL improvement, (re)enabling the tree fold arrow in the same screen, because scrolling through hundreds of animations is a PITA. Can be cherrypicked from second commit. (it's one line)

It would probably be easier to just remove the first commit and keep the second one in your PR instead.

**Review by KoBeWi — APPROVED** [review]

**Comment by YeldhamDev** [comment]
@AThousandShips Changes made.

Conflicts resolved.

**Review by AThousandShips — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Comment by Illauriel** [comment]
Great news! LFG

**Pull request merged**